### PR TITLE
[SPARK-48740][SQL] Catch missing window specification error early

### PR DESCRIPTION
**Pull request by asl3**
### What changes were proposed in this pull request?
Before, aggregate queries containing a window function without a window specification (e.g. `PARTITION BY`) would return a non-descriptive internal error message: 

`org.apache.spark.sql.catalyst.analysis.UnresolvedException: [INTERNAL_ERROR] Invalid call to exprId on unresolved object SQLSTATE: XX000`

This PR catches the user error early and returns a more accurate description of the issue:

`Window specification is not defined in the WINDOW clause for <windowName>. For more information about WINDOW clauses, please refer to '<docroot>/sql-ref-syntax-qry-select-window.html'.`

### Why are the changes needed?
This change produces a more descriptive error message for window functions, improving the Spark user experience. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added unit test in `sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala`

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: common/utils/src/main/resources/error/error-conditions.json [hunk 1]
```diff
@@ -3099,6 +3099,11 @@
     ],
     "sqlState" : "42803"
   },
+  "MISSING_WINDOW_SPECIFICATION" : {
```
Comment (gengliangwang) [inline common/utils/src/main/resources/error/error-conditions.json:3102]: let's also add a sqlState?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala [hunk 2]
```diff
@@ -254,6 +254,14 @@ class AnalysisErrorSuite extends AnalysisTest with DataTypeErrorsBase {
       Array(ExpectedContext("percent_rank(a) FILTER (WHERE c > 1) OVER ()", 7, 50)))
   }
 
+  test("window specification error") {
+    assertAnalysisErrorClass(
+      CatalystSqlParser.parsePlan(s"SELECT avg(a) over (b) FROM TaBlE GROUP BY a, b"),
```
Comment (gengliangwang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisErrorSuite.scala:259]:
```suggestion
      CatalystSqlParser.parsePlan("SELECT avg(a) over (b) FROM TABlE GROUP BY a, b"),
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 3]
```diff
@@ -238,6 +238,13 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
           errorClass = "UNSUPPORTED_FEATURE.OVERWRITE_BY_SUBQUERY",
           messageParameters = Map.empty)
 
+      case agg@Aggregate(_, aggregateExpressions, _) =>
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:241]: shall we put this code next to
```
          case p @ Project(projectList, _) =>
            projectList.foreach(_.transformDownWithPruning(
              _.containsPattern(UNRESOLVED_WINDOW_EXPRESSION)) {
              case UnresolvedWindowExpression(_, windowSpec) =>
                throw QueryCompilationErrors.windowSpecificationNotDefinedError(windowSpec.name)
            })
```

File: common/utils/src/main/resources/error/error-conditions.json [hunk 4]
```diff
@@ -3099,6 +3099,12 @@
...
+    "message" : [
+      "Window specification <windowName> is not defined in the WINDOW clause."
```
Comment (allisonwang-db) [inline common/utils/src/main/resources/error/error-conditions.json:3104]: shall we make the error message more actionable? like "Please make sure it is defined in the WINDOW clause or verify its spelling."

File: common/utils/src/main/resources/error/error-conditions.json [hunk 5]
```diff
@@ -3099,6 +3099,12 @@
     ],
     "sqlState" : "42803"
   },
+  "MISSING_WINDOW_SPECIFICATION" : {
+    "message" : [
+      "Window specification is not defined in the WINDOW clause for <windowName>. Please ensure the WINDOW clause contains PARTITION BY and/or ORDER BY."
```
Comment (gengliangwang) [inline common/utils/src/main/resources/error/error-conditions.json:3104]: The window spec is more than this:
```
windowSpec
    : name=errorCapturingIdentifier                         #windowRef
    | LEFT_PAREN name=errorCapturingIdentifier RIGHT_PAREN  #windowRef
    | LEFT_PAREN
      ( CLUSTER BY partition+=expression (COMMA partition+=expression)*
      | ((PARTITION | DISTRIBUTE) BY partition+=expression (COMMA partition+=expression)*)?
        ((ORDER | SORT) BY sortItem (COMMA sortItem)*)?)
      windowFrame?
      RIGHT_PAREN                                           #windowDef
    ;
```

File: common/utils/src/main/resources/error/error-conditions.json [hunk 6]
```diff
@@ -3099,6 +3099,12 @@
     ],
     "sqlState" : "42803"
   },
+  "MISSING_WINDOW_SPECIFICATION" : {
+    "message" : [
+      "Window specification is not defined in the WINDOW clause for <windowName>. For more information about WINDOW clauses, please refer to the Apache Spark documentation."
```
Comment (gengliangwang) [inline common/utils/src/main/resources/error/error-conditions.json:3104]:
```suggestion
      "Window specification is not defined in the WINDOW clause for <windowName>. For more information about WINDOW clauses, please refer to '<docroot>/sql-ref-syntax-qry-select-window.html'."
```

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
@asl3 please fix the test failures in SQLQueryTestSuite and I will merge this one.

File: sql/core/src/test/resources/sql-tests/analyzer-results/postgreSQL/window_part3.sql.out [hunk 7]
```diff
@@ -336,21 +336,12 @@ org.apache.spark.sql.catalyst.ExtendedAnalysisException
...
-  },
-  "queryContext" : [ {
-    "objectType" : "",
-    "objectName" : "",
-    "startIndex" : 58,
-    "stopIndex" : 67,
-    "fragment" : "GROUP BY 1"
-  } ]
-}
+Project [RANK() OVER (ORDER BY 1 ASC NULLS FIRST ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)#x, count(1)#xL]
```
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/analyzer-results/postgreSQL/window_part3.sql.out:339]: why does it show the plan here? The error message doesn't seem improved in this case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 8]
```diff
@@ -668,11 +677,11 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
...
           case p @ Project(projectList, _) =>
-            projectList.foreach(_.transformDownWithPruning(
-              _.containsPattern(UNRESOLVED_WINDOW_EXPRESSION)) {
-              case UnresolvedWindowExpression(_, windowSpec) =>
-                throw QueryCompilationErrors.windowSpecificationNotDefinedError(windowSpec.name)
-            })
+            checkForUnspecifiedWindow(projectList)
+
+          case agg@Aggregate(_, aggregateExpressions, _) if
+            !PlanHelper.specialExpressionsInUnsupportedOperator(agg).nonEmpty =>
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:683]:
```suggestion
            PlanHelper.specialExpressionsInUnsupportedOperator(agg).isEmpty =>
```

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Pull request closed without merging**